### PR TITLE
[GSOC] 221 Demo notebook for machine translation

### PR DESCRIPTION
**Pull request by LucasSte**
This PR adds a notebook demo for Intel's pre-trained machine translation model, which translates from English to German. The notebook is interactive, so the user can keep typing sentences that get translated instantly.

I'm a recent graduate and I'm interested in working in the project "Add support of Apple M1 to OpenVINO using BNNS library".

### CODE REVIEW

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    "pycharm": {
+     "is_executing": true,
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:36]: @LucasSte please clear the outputs, so CI will run on the notebook.

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Loading and configuring the model\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:96]:
```suggestion
    "## Load and configure the model\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    "The model we use here is available under the `intel/` folder, so we load it and configure its inputs and outputs."
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:97]:
```suggestion
    "The model is now available in the `intel/` folder. Below we load and configure its inputs and outputs."
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Loading tokenizers\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:128]:
```suggestion
    "## Load tokenizers\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    "Before we feed our models with an input sentence, it needs to be transformed into tokens it understands. Likewise, we must transform the output into a sentence we can read.\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:129]:
```suggestion
    "Before providing the model with an input sentence, the input needs to be transformed into tokens the model understands. We must also transform the output into a sentence that we can read.\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+    "\n",
+    "We initialize here the tokenizer for the input `src_tokenizer` and the tokenizer for the output `tgt_tokenizer`."
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:131]:
```suggestion
    "Initialize the tokenizer for the input `src_tokenizer` and the output `tgt_tokenizer`."
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Performing translation\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:162]:
```suggestion
    "## Perform translation\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+    "The following function received a sentence in English and translates it to German."
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:163]:
```suggestion
    "The following function translates a sentence in English to German."
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+   "execution_count": 10,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def translate(sentence):\n",
+    "    # Removes leading and trailing whitespaces\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:177]:
```suggestion
    "    # Remove leading and trailing white spaces\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    "    sentence = sentence.strip()\n",
+    "    assert len(sentence) > 0\n",
+    "    tokens = src_tokenizer.encode(sentence).ids\n",
+    "    # Here we transform the tokenized sentence into the model's input format\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:181]:
```suggestion
    "    # Transform the tokenized sentence into the model's input format\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    "    tokens = [src_tokenizer.token_to_id('<s>')] + tokens + [src_tokenizer.token_to_id('</s>')]\n",
+    "    pad_length = max_tokens - len(tokens)\n",
+    "\n",
+    "    # If the sentence size is smaller the maximum allowed tokens, we fill the remaining tokens with '<pad>'.\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:185]:
```suggestion
    "    # If the sentence size is less than the maximum allowed tokens, fill the remaining tokens with '<pad>'.\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+    "    sentence = sentence.lower().split()\n",
+    "    sentence = \" \".join(key for key, _ in itertools.groupby(sentence))\n",
+    "    return sentence"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Translating sentence\n",
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:212]:
```suggestion
    "## Translate the sentence\n",
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+    "The following function is a basic loop to keep translating sentences."
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:213]:
```suggestion
    "The following function is a basic loop that translates sentences."
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# run_translator() # uncomment this line for a real time translation from your input"
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:249]:
```suggestion
    "# run_translator() # uncomment this line for a real time translation of your input"
```

File: notebooks/215-machine-translation/machine-translation-demo.ipynb [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Translated: mein name ist openvino.\n"
```
Comment (ryanloney) [inline notebooks/215-machine-translation/machine-translation-demo.ipynb:273]: @LucasSte please clear the outputs so the CI will run

**Review by ryanloney — CHANGES_REQUESTED** [review]
@LucasSte some minor copy edits. Thank you for the contribution. This is a fun demo. 
Please also review the other notebooks and follow a similar format. The OMZ model download should not be the first step. Imports should come first, to be consistent with the other notebooks. 

Adrian and developers will provide additional review of the code. In order to get the CI to pass, you must commit a notebook without outputs from executing the cells (needs to be blank) 

Thank you!

**Comment by LucasSte** [comment]
@ryanloney I've updated the README to include the notebook I created. How can I add an image to my table entry?
Can you approve the workflow to check if everything is fine now?

File: notebooks/216-machine-translation/machine-translation-demo.ipynb [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "core = Core()\n",
+    "model = core.read_model('intel/machine-translation-nar-en-de-0002/FP32/machine-translation-nar-en-de-0002.xml')\n",
+    "compiled_model = core.compile_model(model)\n",
+    "infer_request = compiled_model.create_infer_request()\n",
```
Comment (Debskij) [inline notebooks/216-machine-translation/machine-translation-demo.ipynb:113]: We should use single line inference, remove creating infer request

File: notebooks/216-machine-translation/machine-translation-demo.ipynb [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+    "\n",
+    "    # If the sentence size is less than the maximum allowed tokens, fill the remaining tokens with '<pad>'.\n",
+    "    if pad_length > 0:\n",
+    "        tokens = tokens + [src_tokenizer.token_to_id('<pad>')] * pad_length\n",
+    "    assert len(tokens) == max_tokens, \"input sentence is too long\"\n",
+    "    encoded_sentence = np.array(tokens).reshape(1, -1)\n",
+    "\n",
+    "    # Perform inference\n",
+    "    infer_request.infer({input_name: encoded_sentence})\n",
+    "    enc_translated = infer_request.get_tensor(output_name).data[:]\n",
```
Comment (Debskij) [inline notebooks/216-machine-translation/machine-translation-demo.ipynb:193]: Please use single line inference method

**Review by Debskij — CHANGES_REQUESTED** [review]
Hi @LucasSte, thanks for your contribution! Please clear outputs.

File: notebooks/216-machine-translation/machine-translation-demo.ipynb [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def translate(sentence):\n",
```
Comment (zhuo-yoyowz) [inline notebooks/216-machine-translation/machine-translation-demo.ipynb:176]: Please use docstring format and type hints, for all classes, functions, methods. An example could be found here: https://user-images.githubusercontent.com/10940214/163566002-9e552d71-ba99-4d88-afb8-8c439b90aaac.png

File: notebooks/216-machine-translation/machine-translation-demo.ipynb [hunk 19]
```diff
@@ -0,0 +1,304 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Machine translation demo\n",
+    "This demo utilizes Intel's pre-trained model that translates from English to German. More information about the model can be found [here](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/machine-translation-nar-en-de-0002/README.md)."
```
Comment (zhuo-yoyowz) [inline notebooks/216-machine-translation/machine-translation-demo.ipynb:8]: Please introduce a bit more about the model you used in this notebook, to let others know more about the model or any other possible model they could choose for completing the machine translation task from OMZ.

**Review by zhuo-yoyowz — CHANGES_REQUESTED** [review]

**Review by Debskij — APPROVED** [review]

File: notebooks/216-machine-translation/machine-translation-demo.ipynb [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Load tokenizers\n",
+    "Before providing the model with an input sentence, the input needs to be transformed into tokens the model understands. We must also transform the output into a sentence that we can read.\n",
```
Comment (zhuo-yoyowz) [inline notebooks/216-machine-translation/machine-translation-demo.ipynb:103]: Please also include some explanation for what does token mean and why the following files are used here. Suggested comments added here, referring to notebook 213, are "NLP models usually take a list of tokens as standard input. A token is a single word converted to some integer. To provide the proper input, we need the vocabulary for such mapping. Therefore, the following "vocab.json" files are for ......, and the "merge.txt" files are for ......"

**Comment by adrianboguszewski** [comment]
Hi @LucasSte, we're sorry for the long waiting time. We're a small team and were very busy in May and June. We'd like to merge your notebook in July. Are you still interested in doing it?

**Comment by LucasSte** [comment]
> Hi @LucasSte, we're sorry for the long waiting time. We're a small team and were very busy in May and June. We'd like to merge your notebook in July. Are you still interested in doing it?

Hi @adrianboguszewski,
Thanks for the heads up! I'm going to fix these few issues to get it merged.

**Comment by adrianboguszewski** [comment]
Great! Please also change the name to 221-machine-translation (both directory and file) and add the notebook the the main README file.

**Comment by LucasSte** [comment]
> Please put the notebook info in order (after 219) in notebooks/README.md and add it to the main README.md (in root)

Fixed!

File: notebooks/221-machine-translation/221-machine-translation.ipynb [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "## Load tokenizers\n",
+    "\n",
+    "NLP models usually take a list of tokens as standard input. A token is a single word converted to some integer. To provide the proper input, we need the vocabulary for such mapping. We use 'merges.txt' to infor us what sequences of letters form a token. 'vocab.json' specifies the mapping between tokens and integers.\n",
```
Comment (zhuo-yoyowz) [inline notebooks/221-machine-translation/221-machine-translation.ipynb:99]: "We use 'merges.txt' to infor us what sequences of letters form a token." A typo here for "infor"? Do you mean "inform"?

**Review by zhuo-yoyowz — CHANGES_REQUESTED** [review]

File: notebooks/221-machine-translation/221-machine-translation.ipynb (hunk 21 above)
Comment (LucasSte) [reply notebooks/221-machine-translation/221-machine-translation.ipynb:99]: Yes, it should be "inform".

**Review by zhuo-yoyowz — APPROVED** [review]

**Pull request merged**